### ISSUE TITLE
Allow arbitrary precision for generic numeric types

### ISSUE THREAD

**Issue by bendichter** [NeurodataWithoutBorders/pynwb#1631]
### What would you like to see added to PyNWB?

I just got this question from @urut

> We're exporting things with matnwb and running through pynwb validator and it gives error below.    question is, what is meant by 'int' here -- int32? single?

![image (25)](https://user-images.githubusercontent.com/844306/211553339-5d39afec-3dd3-4ef3-9ab6-ec56e7073915.png)

The answer is `int32`, but I think this brings up a good point, and this is a common hurdle users producing NWB files outside of PyNWB. 

IMO, specifying type precision this way has two problems:

1. It's unclear.  I get that single precision is often the default, so `int` can be shorthand for `int32`, but that isn't always the case: MATLAB does not have a default integer precision (`int64(5)` works but `int(5)` does not) and the default precision for floats is double (64). Of course when you are creating a variable you must choose a precision, but when used in a schema to me it's more natural to interpret this an `int` of any precision in this case.
2. It is inflexible. I would also prefer the "any precision" interpretation because it would allow the user to decide what precision they need. Enforcing a higher precision than necessary is wasteful and honestly I don't really see an upside.

### What solution would you like?

My most preferred solution would be to map "int" to "int8", "uint" to "uint8", and "float" to "float8". I'm pretty sure this would not invalidate any currently valid NWB files and would not break any tools.

If there is some reason why keeping single-precision is important, then my next favorite solution would be to change the schema to make the data requirement explicit, changing all instances of "int" to "int32".

If that is too big of a lift, then another solution that would at least move us in the right direction would be to enhance the messages produced by the validator to be explicit about the precision requirement.

### Do you have any interest in helping implement the feature?

Yes.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/.github/CODE_OF_CONDUCT.rst)
- [X] Have you checked the [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured this change was not already [requested](https://github.com/NeurodataWithoutBorders/pynwb/issues)?

**Comment by urut** [comment]
Makes sense to me if the validation for "int" would be "any precision". Often these numbers are small like a TTL so int8 would be sufficient.  Or, at very least, the error should be specific and say what is required (int32) because "int" is ambiguous (and does not exist in Matlab, it has int8, int16, int32, single etc).

**Comment by t-b** [comment]
> Makes sense to me if the validation for "int" would be "any precision".

Yes, that is my preference as well. I can't find the reference right now, the closest I could find was https://github.com/NeurodataWithoutBorders/pynwb/issues/698#issuecomment-435124975. But we have that defined somewhere in the schema.

@bendichter Does the file really contain `int8` data or is just the YAML/JSON datatype `int8`?

**Comment by bendichter** [comment]
Decision from meeting: map "int" to "int8" in the language and in the HDMF library, effectively allowing any precision of integer for anything labeled "int". It will also require a change to the PyNWB validator.
- [ ] change the hdmf schema language docs
- [ ] change the pynwb schema language docs
- [ ] change the hdmf repo to map int to int8, etc.
- [ ] change matnwb to map to int to int8, etc.
- [ ] update pynwb validator

@rly am I missing anything?

**Comment by rly** [comment]
As a note, if the id columns of a `DynamicTable` (type: `ElementIdentifiers`) is initialized to be an int8 dataset and is set to be expandable, then you cannot add more than 128 rows to the table. You would have to delete the dataset and re-create it with a different dtype. Similarly, if a `DynamicTableRegion` is initialized to be an int8 dataset, referenced row indices could not be larger than 128. This is rare and may never come up, but is why @oruebel and I would prefer to keep the dtype of `ElementIdentifiers` and `DynamicTableRegion` to be int32 instead of int8.

I am OK with changing the meaning of "int" to int8, implemented here: https://github.com/hdmf-dev/hdmf-schema-language/pull/38

But note that the only areas where "int" is used nowadays is in `ElementIdentifiers` and `DynamicTableRegion` and if we change the dtype of these types to be int32, then the only places "int" would be observed is in extensions.

**Comment by t-b** [comment]
> I am OK with changing the meaning of "int" to int8, implemented here: https://github.com/hdmf-dev/hdmf-schema-language/pull/38

I don't use `int` in my extensions but won't that break extensions which use `int` as type?

> But note that the only areas where "int" is used nowadays is in ElementIdentifiers and DynamicTableRegion and if we change the dtype of these types to be int32, then the only places "int" would be observed is in extensions.

int32 is still a minimum size right? So we could also use int64 or?

**Comment by rly** [comment]
> I don't use `int` in my extensions but won't that break extensions which use `int` as type?

This change will be set for extensions using the new version of the schema language which will be identified by a comment at the top of the yaml `# nwb-schema-language 3.0.0`. Any extensions lacking that comment are assumed to be from nwb-schema-language 2.0.2 (the current one). 

This will complicate the APIs where they would need to branch behavior depending on the schema language version, but it will improve the system in the long run.

> > But note that the only areas where "int" is used nowadays is in ElementIdentifiers and DynamicTableRegion and if we change the dtype of these types to be int32, then the only places "int" would be observed is in extensions.
> 
> int32 is still a minimum size right? So we could also use int64 or?

Yes, still a minimum size.
